ui/src/app/views/breadcrumb/component.ts

```diff
@@ -0,0 +1,111 @@
+import {Component, OnDestroy, OnInit} from '@angular/core';
+import {ActivatedRoute, NavigationEnd, PRIMARY_OUTLET, Router} from '@angular/router';
+import {Subscription} from 'rxjs/Subscription';
+
+export interface Breadcrumb {
+  label: string;
+  url: string;
+}
+@Component({
+  selector: 'app-breadcrumb',
+  templateUrl: './component.html',
+  styleUrls: ['../../styles/buttons.css',
+    '../../styles/cards.css',
+    '../../styles/headers.css',
+    '../../styles/inputs.css',
+    './component.css']
+})
+export class BreadcrumbComponent implements OnInit, OnDestroy {
+  subscription: Subscription;
+  breadcrumbs: Breadcrumb[];
+  constructor(
+      private activatedRoute: ActivatedRoute,
+      private router: Router) {}
+
+  /**
+   * Generate a breadcrumb using the default label and url. Uses the route's
+   * paramMap to do any necessary variable replacement. For example, if we
+   * have a label value of ':wsid' as defined in a route's breadcrumb, we can
+   * do substitution with the 'wsid' value in the route's paramMap.
+   *
+   * @param {String} label
+   * @param {String} url
+   * @param {ActivatedRoute} route
+   * @returns {Breadcrumb}
+   */
+  private static makeBreadcrumb(
+      label: string,
+      url: string,
+      route: ActivatedRoute
+  ): Breadcrumb {
+    let newLabel = label;
+    // Perform variable substitution in label only if needed.
+    if (newLabel.indexOf(':') >= 0) {
+      try {
+        const paramMap = route.snapshot.paramMap;
+        for (const k of paramMap.keys) {
+          newLabel = newLabel.replace(':' + k, paramMap.get(k));
+        }
+      } catch (e) {
+        console.log(e);
+      }
+    }
+    return {
+      label: newLabel,
+      url: url
+    };
+  }
+
+  ngOnInit() {
+    this.breadcrumbs = this.buildBreadcrumbs(this.activatedRoute.root);
+
+    this.subscription = this.router.events.filter(event => event instanceof NavigationEnd)
+      .subscribe(event => {
+        this.breadcrumbs = this.buildBreadcrumbs(this.activatedRoute.root);
+      });
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  /**
+   * Returns array of Breadcrumb objects that represent the breadcrumb trail.
+   * Derived from current route in conjunction with the overall route structure.
+   *
+   * @param {ActivatedRoute} route
+   * @param {string} url
+   * @param {Breadcrumb[]} breadcrumbs
+   * @returns {Array<Breadcrumb>}
+   */
+  private buildBreadcrumbs(route: ActivatedRoute,
+                           url: string = '',
+                           breadcrumbs: Breadcrumb[] = []): Array<Breadcrumb> {
+
+    const ROUTE_DATA_BREADCRUMB = 'breadcrumb';
+    const children: ActivatedRoute[] = route.children;
+    if (children.length === 0) {
+      return breadcrumbs;
+    }
+    for (const child of children) {
+      if (child.outlet !== PRIMARY_OUTLET) {
```

Comment:
I looked through the docs, but I'm not sure I understand what this is guarding against. Could you add a comment? Is this only set for the currently active child route?